core/src/main/java/org/infinispan/interceptors/impl/PassivationClusteredCacheLoaderInterceptor.java

```diff
@@ -1,24 +1,40 @@
 package org.infinispan.interceptors.impl;
 
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
-import java.util.function.Supplier;
 
 import org.infinispan.commands.FlagAffectedCommand;
 import org.infinispan.container.entries.InternalCacheEntry;
 import org.infinispan.context.InvocationContext;
 import org.infinispan.eviction.impl.ActivationManager;
 import org.infinispan.factories.annotations.Inject;
+import org.infinispan.util.concurrent.CompletionStages;
 import org.infinispan.util.concurrent.DataOperationOrderer;
 
 public class PassivationClusteredCacheLoaderInterceptor<K, V> extends ClusteredCacheLoaderInterceptor<K, V> {
-   @Inject
-   DataOperationOrderer orderer;
+   @Inject DataOperationOrderer orderer;
    @Inject ActivationManager activationManager;
 
    @Override
    public CompletionStage<InternalCacheEntry<K, V>> loadAndStoreInDataContainer(InvocationContext ctx, Object key,
                                                                                 int segment, FlagAffectedCommand cmd) {
-      Supplier<CompletionStage<InternalCacheEntry<K, V>>> supplier = () -> super.loadAndStoreInDataContainer(ctx, key, segment, cmd);
-      return PassivationCacheLoaderInterceptor.handlePassivationLoad(key, segment, orderer, activationManager, supplier);
+       CompletableFuture<DataOperationOrderer.Operation> future = new CompletableFuture<>();
+      CompletionStage<DataOperationOrderer.Operation> delayStage = orderer.orderOn(key, future);
+
+      CompletionStage<InternalCacheEntry<K, V>> retrievalStage;
+      if (delayStage != null && !CompletionStages.isCompletedSuccessfully(delayStage)) {
+         retrievalStage = delayStage.thenCompose(ignore -> super.loadAndStoreInDataContainer(ctx, key, segment, cmd));
+      } else {
+         retrievalStage = super.loadAndStoreInDataContainer(ctx, key, segment, cmd);
+      }
+       if (CompletionStages.isCompletedSuccessfully(retrievalStage)) {
+           InternalCacheEntry<K, V> ice = CompletionStages.join(retrievalStage);
+           PassivationCacheLoaderInterceptor.activateAfterLoad(key, segment, orderer, activationManager, future, ice, null);
+           return retrievalStage;
+       } else {
+           return retrievalStage.whenComplete((value, t) -> {
+               PassivationCacheLoaderInterceptor.activateAfterLoad(key, segment, orderer, activationManager, future, value, t);
+           });
```

Comment:
nit
```suggestion
           return retrievalStage.whenComplete((value, t) -> activateAfterLoad(key, segment, orderer, activationManager, future, value, t));
```

also, you can import the static method `import static org.infinispan.interceptors.impl.PassivationCacheLoaderInterceptor.activateAfterLoad;` to avoid writing the class name and make the lines shorter :)

Reply:
I can do the static import, but I think the line is still too long even without it